parser: remove a bunch of redundant return statements

dchenk opened the pull request:
I’m not sure why func scan had to have so many return statements, but I’m guessing it’s because it’s easy to forget that in Go (unlike in every other language) after a matched case block the break occurs by default. I do understand that it may be preferable to avoid a jump to the end of the switch block if you might as well return, so here I removed only the returns where there really seems to be no advantage to a semi-early return in proportion to the increased size of the function. (Perhaps someone who can pick apart the ASM of the two versions will have something interesting to say.)

dchenk committed:
parser: remove a bunch of redundant return statements

I’m not sure why func scan had to have so many return statements, but I’m guessing it’s because it’s easy to forget that in Go (unlike in every other language) after a matched case block the break occurs by default. I do understand that it may be preferable to avoid a jump to the end of the switch block if you might as well return, so here I removed only the returns where there really seems to be no advantage to a semi-early return in proportion to the increased size of the function. (Perhaps someone who can pick apart the ASM of the two versions will have something interesting to say.)

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/25646)
<!-- Reviewable:end -->

justinj commented:
I think the `return`s were probably so that it was clear that flow ended at that point, since the function is so long with so much nesting it's not especially obvious. I'm kind of ambivalent to this change, as long as we remove all the `return`s, not just some of them.

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---

*[pkg/sql/parser/scan.go, line 364 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25646#-LEFO70d5aha3_RE8HBo:-LEFO70d5aha3_RE8HBp:b-42blo2) ([raw file](https://github.com/cockroachdb/cockroach/blob/420b904c3d5410454244d6ac7ab66d8170214f7d/pkg/sql/parser/scan.go#L364)):*
> ```Go
> 			s.pos++
> 			lval.id = NOT_EQUALS
> 			return
> ```

If we're going to make this change, there's a number of other `return`s to be removed, like this one.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25646#-:-LEFOGlID7XOm-E73O3C:b-vm1q98)*
<!-- Sent from Reviewable.io -->

dchenk commented:
@justinj thanks for the feedback! It looks like I'd need an account on reviewable.io to reply to your comments there, so allow me to post replies here on GitHub :)

On second thought, I think what would actually be better here is to put back a couple of the returns I removed—so that we remove those return statements that literally have no purpose, that is, where a return follows immediately anyway. You'll see what I mean when you look at the changes.


I now made those changes and squashed them into one commit.

justinj commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/parser/scan.go, line 369 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25646#-LEKNdOf65_CiIv5NkYr:-LEKNdOf65_CiIv5NkYs:b-3216r3) ([raw file](https://github.com/cockroachdb/cockroach/blob/195ae3b1c217f2842570820c9018bc8cf4cc1243/pkg/sql/parser/scan.go#L369)):*
> ```Go
> 			s.pos++
> 			lval.id = LESS_EQUALS
> 			return
> ```

Why keep this one and not the one in the `case` below?

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25646)*
<!-- Sent from Reviewable.io -->
